x/epochs/abci.go

```diff
@@ -13,67 +13,43 @@ import (
 // BeginBlocker of epochs module
 func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
 	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
-
-	// Note: epoch length must be more than 2 blocks, as the implementation requires epoch_endblock and epoch_startblock are separate
-	// epoch_startblock(n+1) = epoch_endblock(n) + 1
 	k.IterateEpochInfo(ctx, func(index int64, epochInfo types.EpochInfo) (stop bool) {
-		epochStart := false
-		if !epochInfo.EpochCountingStarted { // epoch counting not started
-			// Should start this epoch timer? (Is StartTime <= ctx.BlockTime)
-			if !epochInfo.StartTime.After(ctx.BlockTime()) {
-				epochStart = true
-				epochInfo.EpochCountingStarted = true
-				epochInfo.CurrentEpoch = 0
-			}
-		} else if epochInfo.CurrentEpochEnded { // epoch ended in last block
-			epochStart = true
-			epochInfo.CurrentEpoch = epochInfo.CurrentEpoch + 1
-		}
+		logger := k.Logger(ctx)
+
+		shouldInitialEpochStart := !epochInfo.EpochCountingStarted && !epochInfo.StartTime.After(ctx.BlockTime())
 
-		if epochStart {
-			epochInfo.CurrentEpochEnded = false
-			epochInfo.CurrentEpochStartTime = ctx.BlockTime()
+		epochEndTime := epochInfo.CurrentEpochStartTime.Add(epochInfo.Duration)
+		shouldEpochStart := ctx.BlockTime().After(epochEndTime) && !epochInfo.StartTime.After(ctx.BlockTime())
```

Comment:
```suggestion
		shouldEpochStart := ctx.BlockTime().After(epochEndTime) && !shouldInitialEpochStart
```

Reply:
But yes we do need !shouldInitialEpochStart at this point as well